File: ros_bazel_installed/ros_apps/oracle.py
#!/usr/bin/env python3

import random

import rclpy
import rclpy.action
import rclpy.duration
import rclpy.node
import rclpy.qos

import ros_apps_msgs.msg
import ros_common_msgs.action
import ros_common_msgs.srv


class Oracle(rclpy.node.Node):

    def __init__(self):
        super().__init__('oracle')
        self._sequence_id = 0
        self._status_pub = self.create_publisher(
            ros_apps_msgs.msg.Status, 'status',
            rclpy.qos.QoSProfile(depth=1))
        self._query_server = self.create_service(
            ros_common_msgs.srv.Query, 'query', self._handle_query)
        self._action_server = rclpy.action.ActionServer(
            self, ros_common_msgs.action.Do, 'do',
            execute_callback=self._handle_rite_action,
            goal_callback=self._handle_action_request,
            cancel_callback=self._handle_cancelled_action,
        )
        self._status_timer = self.create_timer(1.0, self._publish_status)

    def _handle_action_request(self, goal):
        if goal.action != 'rite':
            self.get_logger().warning(
                "Don't know how to " + goal.action)
            return rclpy.action.GoalResponse.REJECT
        return rclpy.action.GoalResponse.ACCEPT

    def _handle_cancelled_action(self, handle):
        return rclpy.action.CancelResponse.ACCEPT

    def _handle_rite_action(self, handle):
        result = ros_common_msgs.action.Do.Result()
        timeout = rclpy.duration.Duration.from_msg(
            handle.request.timeout)
        period = rclpy.duration.Duration.from_msg(
            handle.request.period)
        period = period.nanoseconds / 1e9
        start_time = self.get_clock().now()
        rate = self.create_rate(1 / period)
        while rclpy.ok():
            if handle.is_cancel_requested:
                handle.canceled()
                break
            current_time = self.get_clock().now()
            if current_time - start_time > timeout:
                result.reason = 'timeout'
                handle.abort()
                break
            if bool(random.getrandbits(1)):
                handle.succeed()
                break
            feedback = ros_common_msgs.action.Do.Feedback()
            feedback.message = 'chanting'
            handle.publish_feedback(feedback)
            rate.sleep()
        return result

    def _handle_query(self, request, response):
        if request.query == "how's it going?":
            response.reply = 'all good!'
        else:
            response.reply = "don't know"
        return response

    def _publish_status(self):
        msg = ros_apps_msgs.msg.Status()
        msg.status.sequence_id = self._sequence_id
        self._sequence_id = self._sequence_id + 1
        msg.status.message = 'OK'
        msg.origin = 'oracle'
        self._status_pub.publish(msg)


def main():
    rclpy.init()

    try:
        executor = rclpy.executors.MultiThreadedExecutor()
        rclpy.spin(Oracle(), executor=executor)
    except KeyboardInterrupt:
        pass
    finally:
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_example_bazel_installed/ros2_example_apps/oracle.cc
#include <chrono>
#include <cstdint>
#include <memory>
#include <random>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp/time.hpp>
#include <rclcpp_action/create_server.hpp>
#include <rclcpp_action/server.hpp>

#include "ros2_example_apps_msgs/msg/status.hpp"
#include "ros2_example_common_msgs/action/do.hpp"
#include "ros2_example_common_msgs/srv/query.hpp"

using namespace std::chrono_literals;

namespace ros2_example_apps {

class Oracle : public rclcpp::Node {
  using Status = ros2_example_apps_msgs::msg::Status;
  using Query = ros2_example_common_msgs::srv::Query;
  using Do = ros2_example_common_msgs::action::Do;

 public:
  Oracle() : Node("oracle") {
    using namespace std::placeholders;

    status_pub_ = this->create_publisher<Status>(
        "status", rclcpp::QoS(rclcpp::KeepLast(1)));

    query_server_ = this->create_service<Query>(
        "query", std::bind(&Oracle::handle_query, this, _1, _2));

    action_server_ = rclcpp_action::create_server<Do>(
            this, "do", std::bind(&Oracle::handle_action_goal, this, _1, _2),
            std::bind(&Oracle::handle_cancelled_action, this, _1),
            std::bind(&Oracle::handle_accepted_action, this, _1));

    status_timer_ =
        this->create_wall_timer(1s, std::bind(&Oracle::publish_status, this));
  }

 private:
  rclcpp_action::GoalResponse handle_action_goal(
      const rclcpp_action::GoalUUID&,
      const std::shared_ptr<const Do::Goal> goal) {
    if (goal->action != "rite") {
      RCLCPP_WARN(this->get_logger(), "Don't know how to %s",
                  goal->action.c_str());
      return rclcpp_action::GoalResponse::REJECT;
    }
    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
  }

  using DoGoalHandle = rclcpp_action::ServerGoalHandle<Do>;
  rclcpp_action::CancelResponse handle_cancelled_action(
      const std::shared_ptr<DoGoalHandle>) {
    return rclcpp_action::CancelResponse::ACCEPT;
  }

  void handle_accepted_action(const std::shared_ptr<DoGoalHandle> handle) {
    action_start_time_ = this->get_clock()->now();
    action_loop_ = this->create_wall_timer(
        rclcpp::Duration{handle->get_goal()->period}
            .to_chrono<std::chrono::nanoseconds>(),
        [this, handle]() { this->handle_rite_action(handle); });
  }

  void handle_rite_action(const std::shared_ptr<DoGoalHandle> handle) {
    if (handle->is_canceling()) {
      auto result = std::make_shared<Do::Result>();
      handle->canceled(result);
      action_loop_->cancel();
      return;
    }

    auto current_time = this->get_clock()->now();
    if (current_time - action_start_time_ > handle->get_goal()->timeout) {
      auto result = std::make_shared<Do::Result>();
      result->reason = "timeout";
      handle->abort(result);
      action_loop_->cancel();
      return;
    }

    if (is_rite_complete_(gen_)) {
      auto result = std::make_shared<Do::Result>();
      handle->succeed(result);
      action_loop_->cancel();
      return;
    }

    auto feedback = std::make_shared<Do::Feedback>();
    feedback->message = "chanting";
    handle->publish_feedback(feedback);
  }

  void handle_query(
      const std::shared_ptr<Query::Request> request,
      std::shared_ptr<Query::Response> response)
      const {
    if (request->query == "how's it going?") {
      response->reply = "all good!";
    } else {
      response->reply = "don't know";
    }
  }

  void publish_status() {
    Status msg;
    msg.stamp = this->get_clock()->now();
    msg.status.sequence_id = sequence_id_++;
    msg.status.message = "OK";
    msg.origin = "oracle";
    status_pub_->publish(msg);
  }

  std::random_device rd_;
  std::mt19937 gen_{rd_()};
  std::bernoulli_distribution is_rite_complete_{0.1};

  uint64_t sequence_id_{0u};
  rclcpp::Time action_start_time_;
  std::shared_ptr<rclcpp::TimerBase> action_loop_;
  std::shared_ptr<rclcpp::TimerBase> status_timer_;
  std::shared_ptr<rclcpp::Publisher<Status>> status_pub_;
  std::shared_ptr<rclcpp::Service<Query>> query_server_;
  std::shared_ptr<rclcpp_action::Server<Do>> action_server_;
};

}  // namespace ros_apps

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);

  rclcpp::spin(std::make_shared<ros2_example_apps::Oracle>());

  rclcpp::shutdown();
}
